File: hustack/Ex11_Graph_isomorphism/codes/canon_label.py
from itertools import permutations

def canonical_label(n, edges):
    adj = [[0 for _ in range(n)] for _ in range(n)]
    for u, v in edges:
        adj[u - 1][v - 1] = 1
        adj[v - 1][u - 1] = 1

    best = None
    for perm in permutations(range(n)):
        rep = []
        for i in perm:
            row = []
            for j in perm:
                row.append(str(adj[i][j]))
            rep.append("".join(row))
        represent = "".join(rep)
        if best is None or represent < best:
            best = represent
    return best

def are_isomorphic(n1, n2, edges1, edges2):
    if n1 != n2:
        return False
    label1 = canonical_label(n1, edges1)
    label2 = canonical_label(n2, edges2)
    return label1 == label2

if __name__ == '__main__':
    n1, m1 = list(map(int, input().split()))
    edges1 = []
    for _ in range(m1):
        u, v = list(map(int, input().split()))
        edges1.append((u, v))

    n2, m2 = list(map(int, input().split()))
    edges2 = []
    for _ in range(m2):
        u, v = list(map(int, input().split()))
        edges2.append((u, v))
    
    ans = 1 if are_isomorphic(n1, n2, edges1, edges2) else 0
    print(ans)

File: hustack/Ex11_Graph_isomorphism/codes/canon_label.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

string canonical_label(int n, const vector<pair<int,int>> &edges) {
    vector<vector<int>> adj(n, vector<int>(n, 0));
    for (const auto &e : edges) {
        int u = e.first - 1, v = e.second - 1;
        adj[u][v] = 1;
        adj[v][u] = 1;
    }
    
    vector<int> perm(n);
    for (int i = 0; i < n; i++) 
        perm[i] = i;
    
    string best;
    bool first = true;
    do {
        string rep;
        rep.reserve(n * n);
        for (int i = 0; i < n; i++){
            for (int j = 0; j < n; j++){
                rep.push_back('0' + adj[perm[i]][perm[j]]);
            }
        }
        if (first || rep < best) {
            best = rep;
            first = false;
        }
    } while (next_permutation(perm.begin(), perm.end()));
    
    return best;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n1, m1;
    cin >> n1 >> m1;
    vector<pair<int,int>> edges1(m1);
    for (int i = 0; i < m1; i++){
        int u, v;
        cin >> u >> v;
        edges1[i] = {u, v};
    }
    
    int n2, m2;
    cin >> n2 >> m2;
    vector<pair<int,int>> edges2(m2);
    for (int i = 0; i < m2; i++){
        int u, v;
        cin >> u >> v;
        edges2[i] = {u, v};
    }
    
    if(n1 != n2){
        cout << 0;
        return 0;
    }
    
    string label1 = canonical_label(n1, edges1);
    string label2 = canonical_label(n2, edges2);
    
    cout << (label1 == label2 ? 1 : 0);
    return 0;
}
